contrib/pkg/clusterpool/clusterpool.go

```diff
@@ -0,0 +1,464 @@
+}
+
+func (o *ClusterPoolOptions) getPullSecretSecretName() string {
+	return fmt.Sprintf("%s-pull-secret", o.Name)
+}
+
+func (o *ClusterPoolOptions) addClusterPoolPlatform(cp *hivev1.ClusterPool) {
+	switch o.Cloud {
+	case cloudAWS:
+		cp.Spec.Platform = hivev1.Platform{
+			AWS: &hivev1aws.Platform{
+				CredentialsSecretRef: corev1.LocalObjectReference{
+					Name: o.credsSecretName(),
+				},
+				Region: o.Region,
+			},
+		}
+	case cloudAzure:
+		cp.Spec.Platform = hivev1.Platform{
+			Azure: &hivev1azure.Platform{
+				CredentialsSecretRef: corev1.LocalObjectReference{
+					Name: o.credsSecretName(),
+				},
+				Region:                      o.Region,
+				BaseDomainResourceGroupName: o.AzureBaseDomainResourceGroupName,
+			},
+		}
+	case cloudGCP:
+		cp.Spec.Platform = hivev1.Platform{
+			GCP: &hivev1gcp.Platform{
+				CredentialsSecretRef: corev1.LocalObjectReference{
+					Name: o.credsSecretName(),
+				},
+				Region: o.Region,
+			},
+		}
+	}
+}
+
+func (o *ClusterPoolOptions) credsSecretName() string {
+	if o.createCloudSecret {
+		switch o.Cloud {
+		case cloudAWS:
+			return fmt.Sprintf("%s-aws-creds", o.Name)
+		case cloudAzure:
+			return fmt.Sprintf("%s-azure-creds", o.Name)
+		case cloudGCP:
+			return fmt.Sprintf("%s-gcp-creds", o.Name)
+		}
+		return o.CloudSecretName
+	}
+	return o.CloudSecretName
+}
+
+func (o *ClusterPoolOptions) generateClusterPool() *hivev1.ClusterPool {
+	cp := &hivev1.ClusterPool{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "ClusterPool",
+			APIVersion: hivev1.SchemeGroupVersion.String(),
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        o.Name,
+			Annotations: map[string]string{},
+		},
+		Spec: hivev1.ClusterPoolSpec{
+			BaseDomain: o.BaseDomain,
+			Size:       o.Size,
+		},
+	}
+	if o.PullSecret != "" || o.PullSecretFile != "" {
+		cp.Spec.PullSecretRef = &corev1.LocalObjectReference{Name: o.getPullSecretSecretName()}
+	}
+	if o.ClusterImageSet != "" {
+		cp.Spec.ImageSetRef = hivev1.ClusterImageSetReference{Name: o.ClusterImageSet}
+	}
+	o.addClusterPoolPlatform(cp)
+
+	return cp
+}
+
+func (o *ClusterPoolOptions) generateCredentialsSecret() *corev1.Secret {
+	secret := &corev1.Secret{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Secret",
+			APIVersion: corev1.SchemeGroupVersion.String(),
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      o.credsSecretName(),
+			Namespace: o.Namespace,
+		},
+		Type: corev1.SecretTypeOpaque,
+	}
+	switch o.Cloud {
+	case cloudAWS:
+		secret.StringData = map[string]string{
+			constants.AWSAccessKeyIDSecretKey:     o.awsAccessKeyID,
+			constants.AWSSecretAccessKeySecretKey: o.awsSecretAccessKey,
+		}
+	case cloudAzure:
+		secret.Data = map[string][]byte{
+			constants.AzureCredentialsName: o.azureServicePrincipal,
+		}
+	case cloudGCP:
+		secret.Data = map[string][]byte{
+			constants.GCPCredentialsName: o.gcpServiceAccount,
+		}
+	}
+	return secret
+}
+
+func (o *ClusterPoolOptions) generatePullSecret(pullSecret string) *corev1.Secret {
```

Comment:
Export the `generatePullSecretSecret` from the `Builder` in `pkg/clusterresource`.